# Case 2 - Siemens Advanta

Group W


## 1. Loading the Data

In [1]:
#Importing the necessary libraries

import pandas as pd

The group started by loading the two data files, `Case2_Market data.xlsx` and `Case2_Sales data.csv`.

In [2]:
sales_db = pd.read_csv("datasets\Case2_Sales data.csv", sep=";")
market_db = pd.read_excel("datasets\Case2_Market data.xlsx")

<>:1: SyntaxWarning: invalid escape sequence '\C'
<>:2: SyntaxWarning: invalid escape sequence '\C'
<>:1: SyntaxWarning: invalid escape sequence '\C'
<>:2: SyntaxWarning: invalid escape sequence '\C'
C:\Users\alexr\AppData\Local\Temp\ipykernel_8856\4290426554.py:1: SyntaxWarning: invalid escape sequence '\C'
  sales_db = pd.read_csv("datasets\Case2_Sales data.csv", sep=";")
C:\Users\alexr\AppData\Local\Temp\ipykernel_8856\4290426554.py:2: SyntaxWarning: invalid escape sequence '\C'
  market_db = pd.read_excel("datasets\Case2_Market data.xlsx")


## 2. Data Understanding

### Sales Data
The group decided to take a look a sales_db first.

In [3]:
sales_db

DATE Mapped_GCK Sales_EUR
0     01.10.2018         #1         0
1     02.10.2018         #1         0
2     03.10.2018         #1         0
3     04.10.2018         #1         0
4     05.10.2018         #1         0
...          ...        ...       ...
9797  23.08.2019        #12         0
9798  23.08.2019        #36      1015
9799  12.08.2019        #12         0
9800  28.08.2019         #8  43763,91
9801  27.08.2019         #8         0

[9802 rows x 3 columns]

In [4]:
sales_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9802 entries, 0 to 9801
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   DATE        9802 non-null   object
 1   Mapped_GCK  9802 non-null   object
 2   Sales_EUR   9802 non-null   object
dtypes: object(3)
memory usage: 229.9+ KB


`Sales_EUR` was composed of numeric values yet it was stored as an object. There was also an error in `DATE` which is stored as a object instead of a date. These were the first errors to address so that EDA could be performed.

In [5]:
sales_db['Sales_EUR'] = sales_db['Sales_EUR'].str.replace(',', '.')
sales_db['Sales_EUR'] = sales_db['Sales_EUR'].astype(float)
sales_db['DATE'] = pd.to_datetime(sales_db['DATE'], format='%d.%m.%Y')

In [6]:
sales_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9802 entries, 0 to 9801
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   DATE        9802 non-null   datetime64[ns]
 1   Mapped_GCK  9802 non-null   object        
 2   Sales_EUR   9802 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 229.9+ KB


In [7]:
sales_db.describe()

DATE     Sales_EUR
count                           9802  9.802000e+03
mean   2020-07-19 17:55:48.785961984  2.727857e+05
min              2018-10-01 00:00:00 -5.063812e+05
25%              2019-08-30 00:00:00  0.000000e+00
50%              2020-07-26 12:00:00  0.000000e+00
75%              2021-06-09 00:00:00  0.000000e+00
max              2022-04-30 00:00:00  4.112799e+07
std                              NaN  2.230789e+06

Using the describe method the group found that there was at least one negative value in the `Sales_EUR` column. The group then decided to check for more negative data entries.

In [8]:
negative_sales = sales_db[sales_db['Sales_EUR'] < 0]
negative_sales

DATE Mapped_GCK  Sales_EUR
90   2022-02-02         #6     -54.00
96   2022-04-04        #11    -587.59
102  2021-12-02         #3    -183.50
130  2022-03-02         #6     -93.88
155  2021-05-03         #5    -166.24
...         ...        ...        ...
9333 2019-09-03        #20      -6.60
9466 2018-11-05         #4   -2029.91
9487 2019-04-30        #13     -26.31
9681 2020-07-31         #9     -35.25
9775 2019-03-26         #8   -1000.00

[276 rows x 3 columns]

There were more than 200 rows with negative slaes values, but were there any underlying patterns that could explain these anomalies?

In [9]:
negative_sales['DATE'].dt.year.value_counts()

DATE
2019    88
2021    73
2020    72
2022    32
2018    11
Name: count, dtype: int64

In [10]:
negative_sales['DATE'].dt.month.value_counts()

DATE
4     31
3     30
12    29
11    28
2     26
10    25
1     23
9     21
5     18
7     15
6     15
8     15
Name: count, dtype: int64

In [11]:
negative_sales['DATE'].dt.day.value_counts()

DATE
2     61
4     36
3     35
1     20
6     12
17    10
5     10
12     8
30     8
26     8
7      7
23     5
22     5
27     5
31     4
10     4
21     3
11     3
29     3
16     3
13     3
24     3
8      3
9      3
25     3
14     2
18     2
20     2
19     2
28     2
15     1
Name: count, dtype: int64

The negative values happened at random times, but could they possibly have been connected to a single `Mapped_GCK` (or a group of GCK's)?

In [12]:
negative_sales['Mapped_GCK'].value_counts()

Mapped_GCK
#3     56
#1     51
#5     43
#4     40
#8     19
#9     11
#6     10
#12    10
#36     8
#11     7
#14     7
#16     6
#20     5
#13     3
Name: count, dtype: int64

Much like the dates, it seemed random. The group concluded that there likely wasn't any explanation and this was possibly a data entry error.

### Market Data

In [13]:
market_db

Unnamed: 0  \
0    Index 2010=100 (if not otherwise noted)   
1                                       date   
2                                     2004m2   
3                                     2004m3   
4                                     2004m4   
..                                       ...   
216                                  2021m12   
217                                   2022m1   
218                                   2022m2   
219                                   2022m3   
220                                   2022m4   

                                        China  \
0    Production Index Machinery & Electricals   
1                              MAB_ELE_PRO156   
2                                   16.940704   
3                                   23.711852   
4                                   24.435235   
..                                        ...   
216                                310.763183   
217                                235.956129   
218                                235.956129   
219                                329.413367   
220                                267.373145   

                                     China.1  \
0    Shipments Index Machinery & Electricals   
1                             MAB_ELE_SHP156   
2                                  16.940704   
3                                  23.711852   
4                                  24.435235   
..                                       ...   
216                               310.763183   
217                               235.956129   
218                               235.956129   
219                               329.413367   
220                               267.373145   

                                       France  \
0    Production Index Machinery & Electricals   
1                              MAB_ELE_PRO250   
2                                  112.091273   
3                                  136.327976   
4                                  117.791806   
..                                        ...   
216                                100.565744   
217                                 85.743503   
218                                  90.60354   
219                                107.843548   
220                                  87.69811   

                                    France.1  \
0    Shipments Index Machinery & Electricals   
1                             MAB_ELE_SHP250   
2                                  83.458866   
3                                 106.168192   
4                                  92.007646   
..                                       ...   
216                               134.589504   
217                                108.15632   
218                                117.71577   
219                                136.85872   
220                               116.528738   

                                      Germany  \
0    Production Index Machinery & Electricals   
1                              MAB_ELE_PRO276   
2                                   82.623037   
3                                  100.556582   
4                                   89.653203   
..                                        ...   
216                                118.103281   
217                                  94.55061   
218                                103.987916   
219                                121.308119   
220                                 99.522205   

                                   Germany.1  \
0    Shipments Index Machinery & Electricals   
1                             MAB_ELE_SHP276   
2                                  79.452532   
3                                  97.012918   
4                                  84.932358   
..                                       ...   
216                               149.364286   
217                               120.353403   
218                               129.383676   
219                               151.201314   
220     

First, the group decided to fix the column names so that only one index existed. This was done by joining the country names at the end (or beginning) of the column name.

In [14]:
new_col_names = ['DATE',
                'Production Index Machinery & Electricals China', 'Shipments Index Machinery & Electricals China',
                'Production Index Machinery & Electricals France', 'Shipments Index Machinery & Electricals France',
                'Production Index Machinery & Electricals Germany', 'Shipments Index Machinery & Electricals Germany',
                'Production Index Machinery & Electricals Italy', 'Shipments Index Machinery & Electricals Italy',
                'Production Index Machinery & Electricals Japan', 'Shipments Index Machinery & Electricals Japan',
                'Production Index Machinery & Electricals Switzerland', 'Shipments Index Machinery & Electricals Switzerland',
                'Production Index Machinery & Electricals United Kingdom', 'Shipments Index Machinery & Electricals United Kingdom',
                'Production Index Machinery & Electricals United States', 'Shipments Index Machinery & Electricals United States',
                'Production Index Machinery & Electricals Europe', 'Shipments Index Machinery & Electricals Europe',
                'World: Price of Base Metals', 'World: Price of Energy', 'World: Price of Metals & Minerals', 'World: Price of Natural gas index', 'World: Price of Crude oil, average', 'World: Price of Copper',
                'United States: EUR in LCU',
                'Producer Prices United States: Electrical equipment',
                'Producer Prices United Kingdom: Electrical equipment',
                'Producer Prices Italy: Electrical equipment',
                'Producer Prices France: Electrical equipment',
                'Producer Prices Germany: Electrical equipment',
                'Producer Prices China: Electrical equipment',
                'Production Index United States: Machinery and equipment n.e.c.',
                'Prodcution Index World: Machinery and equipment n.e.c.',
                'Production Index Switzerland: Machinery and equipment n.e.c.',
                'Production Index United Kingdom: Machinery and equipment n.e.c.',
                'Production Index Italy: Machinery and equipment n.e.c.', 
                'Production Index  Japan: Machinery and equipment n.e.c.',
                'Production Index France: Machinery and equipment n.e.c.',
                'Production Index Germany: Machinery and equipment n.e.c.',
                'Production Index United States: Electrical equipment',
                'Production Index World: Electrical equipment',
                'Production Index Switzerland: Electrical equipment',
                'Production Index United Kingdom: Electrical equipment',
                'Production Index Italy: Electrical equipment',
                'Production Index Japan: Electrical equipment',
                'Production Index France: Electrical equipment','Production Index Germany: Electrical equipment']

market_db.columns = new_col_names
market_db.drop(0, inplace=True)

In [15]:
market_db

DATE Production Index Machinery & Electricals China  \
1       date                                 MAB_ELE_PRO156   
2     2004m2                                      16.940704   
3     2004m3                                      23.711852   
4     2004m4                                      24.435235   
5     2004m5                                      23.708115   
..       ...                                            ...   
216  2021m12                                     310.763183   
217   2022m1                                     235.956129   
218   2022m2                                     235.956129   
219   2022m3                                     329.413367   
220   2022m4                                     267.373145   

    Shipments Index Machinery & Electricals China  \
1                                  MAB_ELE_SHP156   
2                                       16.940704   
3                                       23.711852   
4                                       24.435235   
5                                       23.708115   
..                                            ...   
216                                    310.763183   
217                                    235.956129   
218                                    235.956129   
219                                    329.413367   
220                                    267.373145   

    Production Index Machinery & Electricals France  \
1                                    MAB_ELE_PRO250   
2                                        112.091273   
3                                        136.327976   
4                                        117.791806   
5                                        109.002541   
..                                              ...   
216                                      100.565744   
217                                       85.743503   
218                                        90.60354   
219                                      107.843548   
220                                        87.69811   

    Shipments Index Machinery & Electricals France  \
1                                   MAB_ELE_SHP250   
2                                        83.458866   
3                                       106.168192   
4                                        92.007646   
5                                        85.696486   
..                                             ...   
216                                     134.589504   
217                                      108.15632   
218                                      117.71577   
219                                      136.85872   
220                                     116.528738   

    Production Index Machinery & Electricals Germany  \
1                                     MAB_ELE_PRO276   
2                                          82.623037   
3                                         100.556582   
4                                          89.653203   
5                                          86.880571   
..                                               ...   
216                                       118.103281   
217                                         94.55061   
218                                       103.987916   
219                                       121.308119   
220                                        99.522205   

    Shipments Index Machinery & Electricals Germany  \
1                                    MAB_ELE_SHP276   
2                                         79.452532   
3                                         97.012918   
4                                         84.932358   
5                                         82.372794   
..                                              ...   
216                                      149.364286   
217                                      120.353403   
218                                      129.383676   
219                                      151.201314   
220                       

With the new names in order the group decided to create a copy of the data where the codes such as "MAB_ELE_PRO156" where removed. This was done in a separate dataset because at this point the group had no knowledge of the relevance of these codes. To this point the group had theorized that these were probably codes that represented shortened versions of the column names.

In [16]:
market_db1 = market_db.copy()
market_db1.drop(1, inplace=True)
market_db1.iloc[:, 2:48] = market_db1.iloc[:, 2:48].apply(pd.to_numeric, errors='coerce')
market_db1.iloc[:, 2:48] = market_db1.iloc[:, 2:48].astype(float)
market_db1

DATE Production Index Machinery & Electricals China  \
2     2004m2                                      16.940704   
3     2004m3                                      23.711852   
4     2004m4                                      24.435235   
5     2004m5                                      23.708115   
6     2004m6                                      27.009138   
..       ...                                            ...   
216  2021m12                                     310.763183   
217   2022m1                                     235.956129   
218   2022m2                                     235.956129   
219   2022m3                                     329.413367   
220   2022m4                                     267.373145   

    Shipments Index Machinery & Electricals China  \
2                                       16.940704   
3                                       23.711852   
4                                       24.435235   
5                                       23.708115   
6                                       27.009138   
..                                            ...   
216                                    310.763183   
217                                    235.956129   
218                                    235.956129   
219                                    329.413367   
220                                    267.373145   

    Production Index Machinery & Electricals France  \
2                                        112.091273   
3                                        136.327976   
4                                        117.791806   
5                                        109.002541   
6                                        133.785737   
..                                              ...   
216                                      100.565744   
217                                       85.743503   
218                                        90.60354   
219                                      107.843548   
220                                        87.69811   

    Shipments Index Machinery & Electricals France  \
2                                        83.458866   
3                                       106.168192   
4                                        92.007646   
5                                        85.696486   
6                                       106.641482   
..                                             ...   
216                                     134.589504   
217                                      108.15632   
218                                      117.71577   
219                                      136.85872   
220                                     116.528738   

    Production Index Machinery & Electricals Germany  \
2                                          82.623037   
3                                         100.556582   
4                                          89.653203   
5                                          86.880571   
6                                          99.010814   
..                                               ...   
216                                       118.103281   
217                                         94.55061   
218                                       103.987916   
219                                       121.308119   
220                                        99.522205   

    Shipments Index Machinery & Electricals Germany  \
2                                         79.452532   
3                                         97.012918   
4                                         84.932358   
5                                         82.372794   
6                                          95.10874   
..                                              ...   
216                                      149.364286   
217                                      120.353403   
218                                      129.383676   
219                                      151.201314   
220                       

In [17]:
market_db1['DATE'] = pd.to_datetime(market_db1['DATE'].str.strip().str.replace('m', '-'), format='%Y-%m')
market_db1

DATE Production Index Machinery & Electricals China  \
2   2004-02-01                                      16.940704   
3   2004-03-01                                      23.711852   
4   2004-04-01                                      24.435235   
5   2004-05-01                                      23.708115   
6   2004-06-01                                      27.009138   
..         ...                                            ...   
216 2021-12-01                                     310.763183   
217 2022-01-01                                     235.956129   
218 2022-02-01                                     235.956129   
219 2022-03-01                                     329.413367   
220 2022-04-01                                     267.373145   

    Shipments Index Machinery & Electricals China  \
2                                       16.940704   
3                                       23.711852   
4                                       24.435235   
5                                       23.708115   
6                                       27.009138   
..                                            ...   
216                                    310.763183   
217                                    235.956129   
218                                    235.956129   
219                                    329.413367   
220                                    267.373145   

    Production Index Machinery & Electricals France  \
2                                        112.091273   
3                                        136.327976   
4                                        117.791806   
5                                        109.002541   
6                                        133.785737   
..                                              ...   
216                                      100.565744   
217                                       85.743503   
218                                        90.60354   
219                                      107.843548   
220                                        87.69811   

    Shipments Index Machinery & Electricals France  \
2                                        83.458866   
3                                       106.168192   
4                                        92.007646   
5                                        85.696486   
6                                       106.641482   
..                                             ...   
216                                     134.589504   
217                                      108.15632   
218                                      117.71577   
219                                      136.85872   
220                                     116.528738   

    Production Index Machinery & Electricals Germany  \
2                                          82.623037   
3                                         100.556582   
4                                          89.653203   
5                                          86.880571   
6                                          99.010814   
..                                               ...   
216                                       118.103281   
217                                         94.55061   
218                                       103.987916   
219                                       121.308119   
220                                        99.522205   

    Shipments Index Machinery & Electricals Germany  \
2                                         79.452532   
3                                         97.012918   
4                                         84.932358   
5                                         82.372794   
6                                          95.10874   
..                                              ...   
216                                      149.364286   
217                                      120.353403   
218                                      129.383676   
219                                      151.201314   
220 

In [18]:
market_db1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219 entries, 2 to 220
Data columns (total 48 columns):
 #   Column                                                           Non-Null Count  Dtype         
---  ------                                                           --------------  -----         
 0   DATE                                                             219 non-null    datetime64[ns]
 1   Production Index Machinery & Electricals China                   219 non-null    object        
 2   Shipments Index Machinery & Electricals China                    219 non-null    object        
 3   Production Index Machinery & Electricals France                  219 non-null    object        
 4   Shipments Index Machinery & Electricals France                   219 non-null    object        
 5   Production Index Machinery & Electricals Germany                 219 non-null    object        
 6   Shipments Index Machinery & Electricals Germany                  219 non-null    o

In [19]:
nulls = market_db1.isnull().sum()
nulls = nulls[nulls > 0]
nulls = pd.DataFrame(nulls, columns=['nulls'])
nulls['percentage'] = nulls['nulls'] / len(market_db1) * 100
nulls

nulls  percentage
Production Index Machinery & Electricals Switze...      1    0.456621
Shipments Index Machinery & Electricals Switzer...      1    0.456621
Shipments Index Machinery & Electricals United ...     18    8.219178
Shipments Index Machinery & Electricals United ...      1    0.456621
Producer Prices United Kingdom: Electrical equi...     18    8.219178
Producer Prices France: Electrical equipment           35   15.981735
Producer Prices China: Electrical equipment            23   10.502283
Production Index Switzerland: Machinery and equ...      1    0.456621
Production Index World: Electrical equipment           11    5.022831
Production Index Switzerland: Electrical equipment      1    0.456621

Several columns had missing data, but the ones where the msising data percentage was the highest were:
- `Shipments Index Machinery & Electricals United Kingdom` (8%)
- `Producer Prices United Kingdom: Electrical equipment` (8%)
- `Producer Prices France: Electrical equipment` (15%)
- `Producer Prices China: Electrical equipment` (10%)
- `Production Index World: Electrical equipment` (5%).

The group decided to study these columns further to see wich dates had missing values.